app/session.js

```diff
@@ -0,0 +1,111 @@
+var configuration = require('./configuration');
+var authCookie=require('./authenticator.cookie');
+var log = require('./accounter.log');
+
+function parseBody(context,sessionHandler) {
+  var post=require("querystring").parse(context.options.body);
+  var userfield=sessionHandler.userfield || "username";
+  var passfield=sessionHandler.passfield || "password";
+  context.login=post[userfield] || "";
+  context.pw=post[passfield] || "";
+};
+
+function handleSessionRequest(sessionHandler) {
+  var $=this;
+  var context=this.context;
+  switch (this.method) {
+    case "POST":
+      this.readBody(this.context,function(body) {
+        parseBody($.context,sessionHandler);
+        authCookie.ignore($.context,sessionHandler);
+        $.authenticate($.context,function(authenticator) {
+          if (authenticator) {
+            authCookie.set($.context,sessionHandler);
+            if (sessionHandler.forward) {
+              delete context.options.body;
+              $.context.options.method="GET";
+              $.proxyWork($.context);
+            } else {
+              $.sendResponse($.context, 200, 'Authentified');
+            }
+          } else {
+            authCookie.set($.context,sessionHandler);
+            if(sessionHandler.forward) {
+              if (!sessionHandler.preserveCredentials)
+                delete context.options.body;
```

Comment:
This line remove the session credentials if failed and if not explicitly asked. Maybe some particular cases are forgotten in the current state.